libs/libc/pthread/pthread_exit.c

```diff
@@ -0,0 +1,72 @@
+/****************************************************************************
+ * libs/libc/pthread/pthread_exit.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <debug.h>
+#include <sched.h>
+
+#include <nuttx/pthread.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: pthread_exit
+ *
+ * Description:
+ *   Terminate execution of a thread started with pthread_create.
+ *
+ * Input Parameters:
+ *   exit_valie
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+void pthread_exit(FAR void *exit_value)
+{
```

Comment:
pthread_exit() must only be called from user space, especially after this change.  However, I still see several calls to pthread_exit().  In the original code before your change there are these:

    sched/pthread/pthread_create.c:  pthread_exit(exit_status);
    sched/signal/sig_default.c:      pthread_exit(NULL);
    sched/task/task_cancelpt.c:                  pthread_exit(PTHREAD_CANCELED);
    sched/task/task_cancelpt.c:                  pthread_exit(PTHREAD_CANCELED);
    sched/task/task_setcancelstate.c:                  pthread_exit(PTHREAD_CANCELED);
    sched/task/task_setcanceltype.c:              pthread_exit(PTHREAD_CANCELED);

I think this is an error.  We must not call pthread_exit() from within the OS without first dropping the privileges to user.  Otherwise, the problem that we are trying to solve with this PR is not solved.

Reply:
I wonder if any of the calls to pthread_exit() could be replaced with calls to nx_pthread_exit().  That is possible in certain error handling cases where the pthread never ran (and so could not have set up any callbacks).

I wonder if we could move the remaining functions into libs/libc/pthread:

    sched/signal/sig_default.c
    sched/task/task_cancelpt.c
    sched/task/task_setcancelstate.c
    sched/task/task_setcanceltype.c

If we could that might be a simpler solution.

Reply:
The long term objective is to move (almost) all of the pthread logic out of the OS and into the C library.  That is the standard partitioning for example used in GNU/Linux:  Linux implements implements basic native threading but it is GLIBC where all of the pthread interfaces are implemented.  NuttX should follow this same partitioning.

A key step to accomplishing that is to move all cancellation point logic into libs/libc/pthread.  Many pthread APIs are very simple and just built on top of normal OS interfaces but cannot be moved into libs/libc/pthread because the current cancellation point logic works only in supervisor mode.  Moving this to libs/libc/pthread would then enable moving almost all of the other pthread APIs there as well.

Reply:
> pthread_exit() must only be called from user space, especially after this change. However, I still see several calls to pthread_exit(). In the original code before your change there are these:
> 
> ```
> sched/pthread/pthread_create.c:  pthread_exit(exit_status);
> sched/signal/sig_default.c:      pthread_exit(NULL);
> sched/task/task_cancelpt.c:                  pthread_exit(PTHREAD_CANCELED);
> sched/task/task_cancelpt.c:                  pthread_exit(PTHREAD_CANCELED);
> sched/task/task_setcancelstate.c:                  pthread_exit(PTHREAD_CANCELED);
> sched/task/task_setcanceltype.c:              pthread_exit(PTHREAD_CANCELED);
> ```
> 
> I think this is an error. We must not call pthread_exit() from within the OS without first dropping the privileges to user. Otherwise, the problem that we are trying to solve with this PR is not solved.
> 
> Possible solution:
> 
>     1. Add a pointer to pthread_exit to nx_pthread_create().  We have to do this because the address of pthread_exit() will not be known in the PROTECTED and KERNEL builds.  I think you were not seeing the build failure in the PROTECTED/KERNEL builds because cancellation points and default signal actions were not enabled.
> 
>     2. Save the pthread_exit() pointer in the TCB
> 
>     3. Replace calls to pthread_exit() in the OS to up_pthread_exit().
> 
>     4. up_pthread_exit should issue a trap that drops to user mode and calls to the saved pthread_exit() entry point.

I think there is a way to accomplish this without adding a lot of new infrastructure.  I think we could re-use up_pthread_start() to call pthread_exit():

1. Call up_pthread_start() with the pthread entry point equal to NULL.
2. up_pthread_start() will switch to user mode and call the user-space pthread_start() function.
3. pthread_start() would detect the NULL pthread entry point and just call pthread_exit().